### PR TITLE
Delete Doctor By Id

### PR DESCRIPTION
**Pull request by balajisandiparthi2002**
Deletes a doctor by id, if the doctor is not assigned to any patient.

### CODE REVIEW

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java [hunk 1]
```diff
@@ -48,21 +43,27 @@ public ResponseEntity<SuccessResponse<DoctorDto>> createDoctor(@Valid @RequestBo
...
-    public ResponseEntity<?> getDoctorById(@PathVariable(DoctorConstants.PATH_VARIABLE_DOCTOR_ID) String id) {
-        UUID doctorId;
-        try {
-            doctorId = UUID.fromString(id);
-        } catch (IllegalArgumentException illegalArgumentException) {
-            throw new CustomException(messageUtil.getMessage(MessageConstants.INVALID_UUID));
-        }
+    @GetMapping(DoctorConstants.DOCTOR_BY_ID_ENDPOINT)
+    public ResponseEntity<?> getDoctorById(@PathVariable(DoctorConstants.PATH_VARIABLE_DOCTOR_ID) String doctorId) throws CustomException {
         DoctorDto doctorDto = doctorService.getDoctorById(doctorId);
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: Why return type is having "?" ResponseEntity<?> ?
Pls pass the correct response

File: src/main/java/com/theelixrlabs/healthcare/validation/Validator.java [hunk 2]
```diff
@@ -58,4 +61,29 @@ public void validatePatientDTO(PatientDTO patientDTO) throws CustomException {
...
+        if (doctorDto.getLastName().isEmpty()) {
+            throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_LAST_NAME_SHOULD_NOT_BE_EMPTY));
+        } else if (!(doctorDto.getLastName().matches(DoctorConstants.ALPHA_CHARACTERS_REGEX))) {
+            throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_INVALID_LASTNAME));
+        }
+        if (doctorDto.getAadhaarNumber().isEmpty()) {
+            throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_AADHAAR_NUMBER_SHOULD_NOT_BE_EMPTY));
+        } else if (!(doctorDto.getAadhaarNumber().matches(DoctorConstants.AADHAAR_REGEX_PATTERN))) {
+            throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_INVALID_AADHAAR_NUMBER));
+        }
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/validation/Validator.java]: Do you actually need else if (...) ?
Why cant it be simply if block.

File: src/main/resources/messages.properties
Comment (vinodkadam22) [inline src/main/resources/messages.properties:1]: Remove all exclamation marks from this file.
And few messages are having full stop and few are not having. Please follow any one throughout

**Review by vinodkadam22 — CHANGES_REQUESTED** [review]

**Comment by vinodkadam22** [comment]
Resolve your conflicts

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java [hunk 3]
```diff
@@ -48,21 +43,27 @@ public ResponseEntity<SuccessResponse<DoctorDto>> createDoctor(@Valid @RequestBo
...
+    /**
+     * Deletes a doctor by their ID.
+     *
+     * @param doctorId    The ID of the doctor to delete, obtained from the URL path.
+     * @return ResponseEntity containing a SuccessResponse indicating the success status of the deletion operation.
+     * @throws CustomException    If there's an issue with deleting the doctor.
+     */
+    @DeleteMapping(DoctorConstants.DOCTOR_BY_ID_ENDPOINT)
+    public ResponseEntity<SuccessResponse<String>> deleteDoctorById(@PathVariable String doctorId) throws CustomException {
+        return new ResponseEntity<>(new SuccessResponse<>(true, doctorService.deleteDoctorById(doctorId)), HttpStatus.OK);
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: does your success response looks like below
{
success : true,
responseBody : " Doctor with id : {0} deleted Successfully."
}

If yes then do you think responseBody is a correct term for a single string message ?

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java (hunk 1 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: Resolved.

File: src/main/java/com/theelixrlabs/healthcare/validation/Validator.java (hunk 2 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/validation/Validator.java]: Resolved.

File: src/main/resources/messages.properties
Comment (balajisandiparthi2002) [reply src/main/resources/messages.properties:1]: Removed.

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 4]
```diff
@@ -2,41 +2,63 @@
...
     }
 
+    /**
+     * Checks if a doctor is assigned to any patients.
+     *
+     * @param validDoctorId    The unique identifier of the doctor to check.
+     * @return true if the doctor is assigned to at least one patient, false otherwise.
+     */
+    private boolean isDoctorAssignedToPatient(UUID validDoctorId) {
+        List<DoctorPatientAssignmentModel> doctorAssignmentList = doctorPatientAssignmentRepository.findByDoctorId(validDoctorId);
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: instead retrieving list why can't you just use existBy
Also this list may contain assigned and unassigned data as-well 
Please make use of a correct repo method

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 5]
```diff
@@ -87,17 +110,41 @@ public DoctorDto getDoctorById(UUID doctorId) throws CustomException {
...
+        if (!doctorRepository.existsById(validDoctorId))
+            throw new ResourceNotFoundException(messageUtil.getMessage(MessageConstants.DOCTOR_ID_NOT_FOUND));
+
+        // Check if the doctor is assigned to any patients before deletion.
+        if (isDoctorAssignedToPatient(validDoctorId)) {
+            throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_DELETION_FAILED_ASSIGNED_TO_PATIENT));
+        }
+
+        // If no patients are assigned to the doctor, proceed with deletion.
+        doctorRepository.deleteById(validDoctorId);
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java:128]: What about deleting same doctor's un-assignment data from assignment collection ? 
Discuss and update accordingly as per your requirement

**Review by ashutosh-2609 — CHANGES_REQUESTED** [review]

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java (hunk 3 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: Resolved.

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 4 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Resolved.

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 5 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java:128]: If I delete the doctor assignments, then some patients might lose their assignments data.

**Review by tejaswinis-hub — APPROVED** [review]

File: src/main/java/com/theelixrlabs/healthcare/service/PatchDoctorService.java [hunk 6]
```diff
@@ -72,7 +72,7 @@ public DoctorDto patchDoctorById(UUID doctorId, DoctorDto doctorDto) throws Cust
                     .build();
         } else {
             //throw new CustomException(messageUtil.getMessage(MessageConstants.DOCTOR_UNAVAILABLE));
```
Comment (divya-sathyan) [inline src/main/java/com/theelixrlabs/healthcare/service/PatchDoctorService.java]: Remove the unwanted code

File: src/main/java/com/theelixrlabs/healthcare/dao/DoctorPatientAssignmentRepository.java
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/dao/DoctorPatientAssignmentRepository.java:1]: why repository is inside dao ?
Please keep it inside repository

**Review by ashutosh-2609 — CHANGES_REQUESTED** [review]

File: src/main/resources/messages.properties [hunk 7]
```diff
@@ -34,6 +35,6 @@ patient.id.invalid=Invalid Patient Id
 patient.id.not.found=No patient found with specified Id
 patient.last.name.notnull=Patient's Last Name is a mandatory field
 patient.not.found=Patient not found.
-uuid.invalid=UUID not in valid format
+uuid.invalid=Invalid UUID format.
```
Comment (divya-sathyan) [inline src/main/resources/messages.properties]: Rename uuid.invalid ti invalid.uuid

Comment (balajisandiparthi2002) [reply src/main/resources/messages.properties]: Renamed.

File: src/main/java/com/theelixrlabs/healthcare/dao/DoctorPatientAssignmentRepository.java
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/dao/DoctorPatientAssignmentRepository.java:1]: I have moved it.

File: src/main/java/com/theelixrlabs/healthcare/service/PatchDoctorService.java (hunk 6 above)
Comment (balajisandiparthi2002) [reply src/main/java/com/theelixrlabs/healthcare/service/PatchDoctorService.java]: Removed.

**Review by ashutosh-2609 — APPROVED** [review]

**Review by divya-sathyan — APPROVED** [review]

**Review by vinodkadam22 — APPROVED** [review]

**Pull request merged**